### PR TITLE
ledger-configuration: Add tests for the `Configuration` decoding methods. [KVL-1002]

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
- Adds tests.
- Rejects a negative maximum deduplication time (which seemed like an omission).
- Removes a bit of duplication.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by fabiotudone-da — APPROVED** [review]

File: ledger/ledger-configuration/src/test/suite/scala/com/daml/ledger/configuration/ConfigurationSpec.scala [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+        val configuration = Configuration.decode(configurationBytes)
+
+        configuration should be(
+          Left("Missing maximum command time to live")
+        )
+      }
+
+      "rejects a negative transaction latency in the time model" in {
+        val configurationBytes = protobuf.LedgerConfiguration
```
Comment (miklos-da) [inline ledger/ledger-configuration/src/test/suite/scala/com/daml/ledger/configuration/ConfigurationSpec.scala:143]: You may want to rewrite these test cases using table-driven property checks for easier readability (the test cases share most of the code).

File: ledger/ledger-configuration/src/test/suite/scala/com/daml/ledger/configuration/ConfigurationSpec.scala [hunk 2]
```diff
@@ -0,0 +1,244 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
```
Comment (miklos-da) [inline ledger/ledger-configuration/src/test/suite/scala/com/daml/ledger/configuration/ConfigurationSpec.scala:1]: I'd suggest adding a test case for an unknown serialization version as well.

File: ledger/ledger-configuration/src/main/scala/com/daml/ledger/configuration/Configuration.scala [hunk 3]
```diff
@@ -111,6 +45,68 @@ object Configuration {
...
+      .flatMap(decode)
+
+  def decode(config: protobuf.LedgerConfiguration): Either[String, Configuration] =
+    config.getVersion match {
+      case 1 => decodeV1(config)
+      case 2 => decodeV2(config)
+      case v => Left(s"Unknown version: $v")
+    }
+
+  def decodeV1(config: protobuf.LedgerConfiguration): Either[String, Configuration] =
```
Comment (miklos-da) [inline ledger/ledger-configuration/src/main/scala/com/daml/ledger/configuration/Configuration.scala:60]:
```suggestion
  private def decodeV1(config: protobuf.LedgerConfiguration): Either[String, Configuration] =
```

File: ledger/ledger-configuration/src/main/scala/com/daml/ledger/configuration/Configuration.scala [hunk 4]
```diff
@@ -111,6 +45,68 @@ object Configuration {
...
+        }
+    } yield {
+      Configuration(
+        generation = config.getGeneration,
+        timeModel = tm,
+        maxDeduplicationTime = Duration.ofDays(1),
+      )
+    }
+
+  def decodeV2(config: protobuf.LedgerConfiguration): Either[String, Configuration] =
```
Comment (miklos-da) [inline ledger/ledger-configuration/src/main/scala/com/daml/ledger/configuration/Configuration.scala:76]:
```suggestion
  private def decodeV2(config: protobuf.LedgerConfiguration): Either[String, Configuration] =
```

**Review by miklos-da — APPROVED** [review]

**Review by hubert-da — APPROVED** [review]
+1 for suggestions from Miklos

File: ledger/ledger-configuration/src/test/suite/scala/com/daml/ledger/configuration/ConfigurationSpec.scala (hunk 1 above)
Comment (SamirTalwar-DA) [reply ledger/ledger-configuration/src/test/suite/scala/com/daml/ledger/configuration/ConfigurationSpec.scala]: Good call. Done for the error cases.

File: ledger/ledger-configuration/src/test/suite/scala/com/daml/ledger/configuration/ConfigurationSpec.scala (hunk 2 above)
Comment (SamirTalwar-DA) [reply ledger/ledger-configuration/src/test/suite/scala/com/daml/ledger/configuration/ConfigurationSpec.scala:1]: Good idea. I added one for a version 3, which should also prompt anyone adding a v3 to also add tests. :slightly_smiling_face:

**Pull request merged**